<a href="https://colab.research.google.com/github/francesco-zatto/dia-project/blob/master/car_emissions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [166]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [167]:
import os.path
from urllib.request import urlretrieve

DATASETS = {
    "car_emissions_spain2022.csv": "https://raw.githubusercontent.com/pietroolivi/dia-datasets/main/car_emissions_spain2022.csv",
    "car_emissions_canada2014.csv": "https://raw.githubusercontent.com/pietroolivi/dia-datasets/main/car_emissions_canada2014/car_emissions_canada2014.csv",
    "car_emissions_uk2013.csv": "https://raw.githubusercontent.com/pietroolivi/dia-datasets/main/car_emissions_uk2013.csv"
}

for filename, dataset in DATASETS.items():
    if not os.path.exists(filename):
      urlretrieve(dataset, filename)

In [168]:
spain_emissions = pd.read_csv("car_emissions_spain2022.csv")
spain_emissions.head(5)

id    make                          model               market_segment  \
0  551266  AIWAYS  AIWAYS U5 MAS861-WVTA/2WB/FL4   Berlinas-Familiares Medios   
1  579684  AIWAYS                AIWAYS U6 PRIME  Berlinas-Familiares Grandes   
2  464453    ALKE     ALKE ATX 310 E plomo acido        Chasis-Cabina Pequeño   
3  464454    ALKE     ALKE ATX 320 E plomo acido        Chasis-Cabina Pequeño   
4  464455    ALKE     ALKE ATX 330 E plomo acido        Chasis-Cabina Pequeño   

        engine_type  consumption_min_l_100km  consumption_max_l_100km  \
0  Eléctricos puros                      NaN                      NaN   
1  Eléctricos puros                      NaN                      NaN   
2  Eléctricos puros                      NaN                      NaN   
3  Eléctricos puros                      NaN                      NaN   
4  Eléctricos puros                      NaN                      NaN   

   emissions_min_gCO2_km  emissions_max_gCO2_km transmission  ...  \
0                    NaN                    NaN            A  ...   
1                    NaN                    NaN            A  ...   
2                    NaN                    NaN           SC  ...   
3                    NaN                    NaN           SC  ...   
4                    NaN                    NaN           SC  ...   

   avg_wltp_emissions_gCO2_km  length_mm  width_mm  height_mm  \
0                         0.0       4680      1865       1700   
1                         0.0       4805      1880       1641   
2                         0.0       3030      1500       2000   
3                         0.0       3850      1500       2000   
4                         0.0       3850      1500       2000   

   gross_vehicle_weight_rating_kg  total_seating  fuel_economy_index  \
0                            2135              5   Sin clasificación   
1                            2290              5   Sin clasificación   
2                            1510              2   Sin clasificación   
3                            1510              2   Sin clasificación   
4                            1510              2   Sin clasificación   

   type_hybrid  electric_consumption_kwh_100km  battery_capacity_kwh  
0          NaN                           15.75                  63.0  
1          NaN                           16.00                  63.0  
2          NaN                           11.00                  10.0  
3          NaN                           11.00                  10.0  
4          NaN                           11.00                  10.0  

[5 rows x 26 columns]

In [169]:
canada_emissions = pd.read_csv("car_emissions_canada2014.csv")
canada_emissions.head(5)

_id  Model year   Make         Model Vehicle class  Engine size (L)  \
0    1        1995  Acura       Integra    Subcompact              1.8   
1    2        1995  Acura       Integra    Subcompact              1.8   
2    3        1995  Acura  Integra GS-R    Subcompact              1.8   
3    4        1995  Acura        Legend       Compact              3.2   
4    5        1995  Acura  Legend Coupe       Compact              3.2   

   Cylinders Transmission Fuel type  City (L/100 km)  Highway (L/100 km)  \
0          4           A4         X             10.2                 7.0   
1          4           M5         X              9.6                 7.0   
2          4           M5         Z              9.4                 7.0   
3          6           A4         Z             12.6                 8.9   
4          6           A4         Z             13.0                 9.3   

   Combined (L/100 km)  Combined (mpg)  CO2 emissions (g/km)  
0                  8.8              32                   202  
1                  8.4              34                   193  
2                  8.3              34                   191  
3                 10.9              26                   251  
4                 11.3              25                   260

In [170]:
uk_emissions = pd.read_csv("car_emissions_uk2013.csv")
uk_emissions.head(5)

C:\Users\franc\AppData\Local\Temp\ipykernel_8872\3678210660.py:1: DtypeWarning: Columns (6,30) have mixed types. Specify dtype option on import or set low_memory=False.
  uk_emissions = pd.read_csv("car_emissions_uk2013.csv")


file  year manufacturer      model         description  \
0  DatapartC_july2000.csv  2000   Alfa Romeo  145 Range  1.6 Twin Spark 16v   
1  DatapartC_july2000.csv  2000   Alfa Romeo  145 Range  1.8 Twin Spark 16v   
2  DatapartC_july2000.csv  2000   Alfa Romeo  145 Range          Cloverleaf   
3  DatapartC_july2000.csv  2000   Alfa Romeo  146 Range  1.6 Twin Spark 16v   
4  DatapartC_july2000.csv  2000   Alfa Romeo  146 Range  1.8 Twin Spark 16v   

   euro_standard tax_band transmission transmission_type  engine_capacity  \
0              2      NaN           M5            Manual           1598.0   
1              2      NaN           M5            Manual           1747.0   
2              2      NaN           M5            Manual           1970.0   
3              2      NaN           M5            Manual           1598.0   
4              2      NaN           M5            Manual           1747.0   

   ... nox_emissions  thc_nox_emissions  particulates_emissions  \
0  ...           NaN                NaN                     NaN   
1  ...           NaN                NaN                     NaN   
2  ...           NaN                NaN                     NaN   
3  ...           NaN                NaN                     NaN   
4  ...           NaN                NaN                     NaN   

   fuel_cost_12000_miles  fuel_cost_6000_miles  standard_12_months  \
0                    NaN                 618.0                 NaN   
1                    NaN                 633.0                 NaN   
2                    NaN                 663.0                 NaN   
3                    NaN                 626.0                 NaN   
4                    NaN                 633.0                 NaN   

   standard_6_months  first_year_12_months  first_year_6_months  \
0                NaN                   NaN                  NaN   
1                NaN                   NaN                  NaN   
2                NaN                   NaN                  NaN   
3                NaN                   NaN                  NaN   
4                NaN                   NaN                  NaN   

   date_of_change  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  

[5 rows x 31 columns]

# **Esplorazione dei singoli dataset**

## Esplorazione del dataset `spain_emissions`

### Rimozione delle righe riguardanti veicoli elettrici

All'interno del dataset proveniente dal ministero spagnolo sono presenti anche dati riguardanti macchine elettriche. Questa tipologia di dato non è interessante per l'obiettivo del modello, ovvero la quantità di CO2 emessa.

La cella seguente analizza i valori delle righe che hanno `engine_type` "Eléctricos puros", estraendo solo le colonne sul minimo e massimo di consumo di litri di carburante e sul minimio e massimo di emissioni di CO2 per km percorso.

In [171]:
columns = ["emissions_min_gCO2_km", "emissions_max_gCO2_km"]
spain_emissions.loc[spain_emissions["engine_type"] == "Eléctricos puros", columns].isna().all()

emissions_min_gCO2_km    True
emissions_max_gCO2_km    True
dtype: bool

Come era previdibile, tutti i valori delle emissioni di CO2 per km nelle righe riguardanti i veicoli elettrici sono mancanti, quindi le rispettive righe possono essere eliminate, poichè non utili per l'addestramento del modello, il cui obiettivo è la previsione dell'emissione di C02 per veicoli a motore termico.

In [172]:
electric_cars_index = spain_emissions.loc[spain_emissions["engine_type"] == "Eléctricos puros"].index
spain_emissions = spain_emissions.drop(electric_cars_index)
spain_emissions["engine_type"].unique()

array(['Gasolina', 'Gasóleo', 'Híbridos enchufables',
       'Híbridos de gasolina', 'Gas natural', 'Híbridos de gasóleo',
       'Gases licuados del petróleo (GLP)', 'Pila de combustible',
       'Autonomía extendida'], dtype=object)

Come si può notare dall'output della cella precedente, la colonna `engine_type` non contiene il valore "Eléctricos puros".

### Traduzione spagnolo-inglese 
Un'altra elaborazione necessaria è una traduzione dallo spagnolo all'inglese sia per questa colonna, che per la colonna della classificazione delle auto, ovvero `market_segment`.

Definisco una funzione per sostituire i valori di una colonna di un dataframe con dei nuovi valori passati come argomento, in questo caso la rispettiva traduzione in inglese, tramite una funzione di mapping fornita da pandas.

In [173]:
def get_column_mapped_values(column, new_values):
    mapping = dict(zip(column.unique(), new_values))
    return column.map(mapping)

In [174]:
#Traduzione della colonna engine_type
spanish_engine_type = spain_emissions["engine_type"].unique()
english_engine_type = ["Petrol", "Diesel", "Plug-in hybrid", "Petrol hybrid", "Natural gas", 
                       "Diesel Hybrid", "Liquefied petroleum gas(LPG)", "Fuel cell", "Extended range"]

spain_emissions["engine_type"] = get_column_mapped_values(spain_emissions["engine_type"], english_engine_type)

Per la traduzione della colonna `market_segment` si fa riferimento al dataset canadese, il quale contiene una colonna per la classificazione dei veicoli, `Vehicle_class`.

In [175]:
canada_emissions["Vehicle class"].unique()

array(['Subcompact', 'Compact', 'Two-seater', 'Station wagon: Small',
       'Minicompact', 'Mid-size', 'Full-size', 'Station wagon: Mid-size',
       'Van: Cargo', 'Van: Passenger', 'Pickup truck: Standard',
       'Sport utility vehicle', 'Minivan', 'Pickup truck: Small',
       'Special purpose vehicle', 'Sport utility vehicle: Small',
       'Sport utility vehicle: Standard'], dtype=object)

In [176]:
#Traduzione della colonna market_segment
english_market_segment = ["Minicompact", "Compact", "Small off-road", "Mid-size", "Sport utility vehicle", "Full-size", 
                          "Mid-size off-road", "Full-size off-road", "Minivan", "Luxury", "Minivan", "Van: Passenger"] \
                        + (["Van: Cargo"] * 2) + (["Lorry"] * 7)

spain_emissions["market_segment"] = get_column_mapped_values(spain_emissions["market_segment"], english_market_segment)

### Calcolo del volume del motore in cm<sup>3</sup>
Il dataset contiene tre colonne che rispettivamente indicano le tre dimensioni del motore, ovvero lunghezza, larghezza e altezza in millimetri. Queste tre colonne potrebbero essere delle feature molto importanti per il modello. Quindi è necessario mantenerle nel dataset per motivi di compatibilità con gli altri dataset in una nuova colonna `engine_capacity_cm3`, il volume in centimetri cubi, dato dal prodotto delle tre colonne e convertito successivamente in cm<sup>3</sup>.


In [177]:
spain_emissions["engine_capacity_cm3"] = spain_emissions["length_mm"] * spain_emissions["width_mm"] * spain_emissions["height_mm"]
spain_emissions["engine_capacity_cm3"] /= 1000

### Rimozione delle colonne contenenti solo valori `NaN`
Nel dataset dell'emissioni di CO<sub>2</sub> proveniente dalla Spagna la colonne `type_hybrid`, `electric_consumption_kwh_100km`, `battery_capacity_kwh ` contengono solo valori `NaN`.

In [178]:
all_nan_columns = ["type_hybrid", "electric_consumption_kwh_100km", "battery_capacity_kwh"]
spain_emissions[all_nan_columns].isna().values.all()

True

Quindi possono essere rimosse dal dataset, poiché non sono feature interessanti ed utili per l'addestramento del modello.

In [179]:
spain_emissions = spain_emissions.drop(all_nan_columns, axis=1)

### Calcolo medie del consumo di carburante e delle emissioni di CO<sub>2</sub>
Sia per il consumo di carburante che per le emissioni di CO<sub>2</sub> sono presenti i valori di minimo, di massimo e la media
WLTP (<i>Worldwide harmonized Light vehicles Test Procedure</i>), ovvero le misure ottenute dal test standard a livello mondiale per il controllo delle emissioni e dei consumi dei veicoli.
Da questi tre valori ne verrà calcolata la media così da avere una sola colonna per i consumi di carburante e una sola colonna per i consumi di CO<sub>2</sub>, la variabile target da predire.

Prima di poter precedere è necessario controllare che in tutte le righe del dataframe sia presente almeno uno dei tre valori richiesti e che sia diverso da 0, sia per calcolare la media del consumo che delle emissioni.
<br>Nella cella successiva si può vedere esattamente in quante righe sono mancanti questi valori necessari.

In [180]:
def isnaOrIsZero(dataframe):
    return dataframe.isna() | dataframe.eq(0)

In [181]:
consumption_columns = ["consumption_min_l_100km", "consumption_max_l_100km", "avg_wltp_consumption_l_100km"]
emissions_columns = ["emissions_min_gCO2_km", "emissions_max_gCO2_km", "avg_wltp_emissions_gCO2_km"]

rows_no_consumption = isnaOrIsZero(spain_emissions[consumption_columns]).all(axis=1)
number_rows_missing_consumption = rows_no_consumption.sum()
print("Rows with no consumption_l_100km:", number_rows_missing_consumption)

rows_no_emissions = isnaOrIsZero(spain_emissions[emissions_columns]).all(axis=1)
number_rows_missing_emissions = rows_no_emissions.sum()
print("Rows with no emissions_gCO2_km:", number_rows_missing_emissions)

rows_no_emissions_consumptions = (rows_no_consumption & rows_no_emissions)
number_rows_missing_emissions_consumptions = rows_no_emissions_consumptions.sum()
print("Rows with no emissions_gCO2_km and no consumption_l_km:", number_rows_missing_emissions_consumptions)

Rows with no consumption_l_100km: 1560
Rows with no emissions_gCO2_km: 1562
Rows with no emissions_gCO2_km and no consumption_l_km: 1560


Come riportato nell'output della cella precedente, ci sono <strong>1560</strong> righe senza alcun valore di consumo di carburante, <strong>1562</strong> righe senza alcun valore di emissioni di CO<sub>2</sub> e <strong>1560</strong> in cui è assente sia il consumo di carburante che le emissioni di CO<sub>2</sub>. <br> 
Nella cella successiva vengono calcolate le medie di carburante ed emissioni di ciascuna riga e vengono inserite in due nuove colonne `consumption_l_100km`, `emissions_gCO2_km`. <br>

In [182]:
mean_consumption = "consumption_l_100km"
mean_emissions = "emissions_gCO2_km"
spain_emissions[mean_consumption] = spain_emissions[consumption_columns].mean(axis=1)
spain_emissions[mean_emissions] = spain_emissions[emissions_columns].mean(axis=1)

In [183]:
spain_emissions = spain_emissions.drop(consumption_columns + emissions_columns, axis=1)

Quindi vengono sostituiti i valori mancanti e i valori posti a 0, poichè considerati anch'essi mancanti, siccome le categorie di veicoli considerate dovrebbero presentare dei consumi di carburante e delle emissioni di CO<sub>2</sub> positivi. <br>
La cella successiva mostra le categorie di veicoli in cui sono assenti sia i consumi di carburante che le emissioni. Le due categorie con più occorrenze sono i "Van: Cargo" e "Lorry", due mezzi molto pesanti e con valori differenti, quindi sarebbe più corretto riempire i valori assenti o posti a 0, con i mediani delle due rispettive categorie. Mentre per le restanti categorie si può utilizzare direttamente il mediano di tutta la colonna.

In [184]:
most_missing_market_segments = spain_emissions.loc[rows_no_emissions_consumptions, "market_segment"].value_counts()[:2]
most_missing_market_segments

Van: Cargo    1202
Lorry          327
Name: market_segment, dtype: int64

In [208]:
def fillnaZeroes(market_segment, columns):
    vehicles = spain_emissions.loc[spain_emissions["market_segment"] == market_segment, columns].replace(0, np.nan)
    vehicles = vehicles.fillna(vehicles.median())
    return vehicles

In [209]:
market_segments = ["Van: Cargo", "Lorry"]
columns = [mean_consumption, mean_emissions]
for market_segment in market_segments:
    spain_emissions.loc[spain_emissions["market_segment"] == market_segment, columns] = fillnaZeroes(market_segment, columns)

### Grafici 
Questa sezione contiene alcuni grafici che evidenziano alcuni aspetti del dataframe risulante `spain_emissions`.

In [165]:
spain_emissions["emissions_max_gCO2_km"].plot.hist();

KeyError: 'emissions_max_gCO2_km'

In [ ]:
spain_emissions["consumption_max_l_100km"].plot.hist();

In [ ]:
spain_emissions.plot.scatter("consumption_max_l_100km", "emissions_max_gCO2_km");

# **Omogeneizzazione ed unione dei dati**
Avendo reperito tre set di dati ciascuno da una fonte differente, sebbene questi concernano il medesimo dominio e condividano pertanto una considerevole quantità di attributi,

Potremmo essere erroneamente portati a compiere un troncamento delle colonne, conservando esclusivamente quelle appartenenti all'intersezione delle tre tabelle, tuttavia così facendo andremmo a trascurare il caso-limite in cui le features eliminate dovessero essere le uniche legate, attraverso un relazione non randomica, a quella indagata. Una soluzione più scrupolosa consisterebbe, al contrario, nell'unione degli attributi, colmando opportunamente le celle vuote formatesi. In effetti, come visto a lezione, i valori NAN possono essere sostituiti, tra gli altri, da:
- media
- moda
- mediana

In aggiunta, le stesse variabili in comune, malgrado facciano riferimento allo stesso concetto, si manifestano talvolta sotto denominazioni diverse, ed a loro volta i valori corrispondenti da queste assunti potrebbero avere formato dissimile (e.g. capitalizzazione delle lettere nel caso testuale o numero di cifre significative in quello numerico).





